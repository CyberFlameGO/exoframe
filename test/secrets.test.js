/* eslint-env jest */
// mock config for testing
jest.mock('../src/config', () => require('./__mocks__/config'));

// npm packages
const nock = require('nock');
const sinon = require('sinon');
const inquirer = require('inquirer');

// our packages
const {handler: secrets} = require('../src/commands/secrets');
const cfg = require('../src/config');

const testSecret = {
  secretName: 'test',
  secretValue: '12345',
};

// test generation
test('Should create new secret', done => {
  // handle correct request
  const secretServer = nock('http://localhost:8080')
    .post('/secrets')
    .reply(200, {name: testSecret.secretName, value: testSecret.secretValue});
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // stup inquirer answers
  sinon.stub(inquirer, 'prompt').callsFake(() => Promise.resolve(testSecret));
  // execute login
  secrets({}).then(() => {
    // make sure log in was successful
    // check that server was called
    expect(secretServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore console
    console.log.restore();
    // restore inquirer
    inquirer.prompt.restore();
    // tear down nock
    secretServer.done();
    done();
  });
});

// test non-interactive generation
test('Should create new secret non-interactively', done => {
  // handle correct request
  const secretServer = nock('http://localhost:8080')
    .post('/secrets')
    .reply(200, {name: testSecret.secretName, value: testSecret.secretValue});
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // execute login
  secrets({cmd: 'new', name: testSecret.secretName, value: testSecret.secretValue}).then(() => {
    // make sure log in was successful
    // check that server was called
    expect(secretServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore console
    console.log.restore();
    // tear down nock
    secretServer.done();
    done();
  });
});

// test list
test('Should list secrets', done => {
  const createDate = new Date(2017, 1, 1, 1, 1, 1, 1);
  // handle correct request
  const secretsServer = nock('http://localhost:8080')
    .get('/secrets')
    .reply(200, {secrets: [{name: testSecret.secretName, meta: {created: createDate}}]});
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // execute login
  secrets({cmd: 'ls'}).then(() => {
    // make sure log in was successful
    // check that server was called
    expect(secretsServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args.map(lines => lines.map(l => l.replace(createDate.toLocaleString(), '')))).toMatchSnapshot();
    // restore console
    console.log.restore();
    // tear down nock
    secretsServer.done();
    done();
  });
});

// test getting
test('Should get secret value', done => {
  const createDate = new Date(2018, 1, 1, 1, 1, 1, 1);
  // handle correct request
  const secretGetServer = nock('http://localhost:8080')
    .get('/secrets')
    .reply(200, {secrets: [{name: testSecret.secretName, meta: {created: createDate}}]});
  // handle correct request
  const secretServer = nock('http://localhost:8080')
    .get(`/secrets/${testSecret.secretName}`)
    .reply(200, {secret: {...testSecret, meta: {created: createDate}}});
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // stup inquirer answers
  sinon
    .stub(inquirer, 'prompt')
    .onFirstCall()
    .callsFake(() => Promise.resolve({selectedSecret: testSecret.secretName}))
    .onSecondCall()
    .callsFake(() => Promise.resolve({doGet: true}));
  // execute login
  secrets({cmd: 'get'}).then(() => {
    // make sure log in was successful
    // check that server was called
    expect(secretGetServer.isDone()).toBeTruthy();
    expect(secretServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore console
    console.log.restore();
    // restore inquirer
    inquirer.prompt.restore();
    // tear down nock
    secretGetServer.done();
    secretServer.done();
    done();
  });
});

test('Should list zero secrets', done => {
  // handle correct request
  const secretsServer = nock('http://localhost:8080').get('/secrets').reply(200, {secrets: []});
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // execute login
  secrets({cmd: 'ls'}).then(() => {
    // make sure log in was successful
    // check that server was called
    expect(secretsServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore console
    console.log.restore();
    // tear down nock
    secretsServer.done();
    done();
  });
});

// test removal
test('Should remove secret', done => {
  const createDate = new Date();
  // handle correct request
  const secretGetServer = nock('http://localhost:8080')
    .get('/secrets')
    .reply(200, {secrets: [{name: testSecret.secretName, meta: {created: createDate}}]});
  // handle correct request
  const secretServer = nock('http://localhost:8080').delete('/secrets').reply(204, '');
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // stup inquirer answers
  sinon.stub(inquirer, 'prompt').callsFake(() => Promise.resolve({selectedSecret: testSecret.secretName}));
  // execute login
  secrets({cmd: 'rm'}).then(() => {
    // make sure log in was successful
    // check that server was called
    expect(secretGetServer.isDone()).toBeTruthy();
    expect(secretServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore console
    console.log.restore();
    // restore inquirer
    inquirer.prompt.restore();
    // tear down nock
    secretGetServer.done();
    secretServer.done();
    done();
  });
});

// test deauth
test('Should deauth on 401 on creation', done => {
  // save current config state
  cfg.__save('token');
  // handle correct request
  const secretServer = nock('http://localhost:8080').post('/secrets').reply(401);
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // stup inquirer answers
  sinon.stub(inquirer, 'prompt').callsFake(() => Promise.resolve({secretName: 'test'}));
  // execute login
  secrets({}).then(() => {
    // make sure log in was successful
    // check that server was called
    expect(secretServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore console
    console.log.restore();
    // restore inquirer
    inquirer.prompt.restore();
    // tear down nock
    secretServer.done();
    done();
  });
});

test('Should deauth on 401 on list', done => {
  // restore config with auth
  cfg.__restore('token');
  // handle correct request
  const secretServer = nock('http://localhost:8080').get('/secrets').reply(401);
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // stup inquirer answers
  sinon.stub(inquirer, 'prompt').callsFake(() => Promise.resolve({secretName: 'test'}));
  // execute login
  secrets({cmd: 'ls'}).then(() => {
    // make sure log in was successful
    // check that server was called
    expect(secretServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore console
    console.log.restore();
    // restore inquirer
    inquirer.prompt.restore();
    // tear down nock
    secretServer.done();
    done();
  });
});
