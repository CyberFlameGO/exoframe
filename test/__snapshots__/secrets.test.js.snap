// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create new secret 1`] = `
Array [
  Array [
    "[1mGenerating new deployment secret for:[22m",
    "http://localhost:8080",
  ],
  Array [
    "[1mNew secret generated:[22m",
  ],
  Array [
    "",
  ],
  Array [
    "Name: test",
  ],
  Array [
    "Value: 12345",
  ],
  Array [
    "",
  ],
  Array [
    "[32mDONE![39m",
  ],
]
`;

exports[`Should create new secret non-interactively 1`] = `
Array [
  Array [
    "[1mGenerating new deployment secret for:[22m",
    "http://localhost:8080",
  ],
  Array [
    "[1mNew secret generated:[22m",
  ],
  Array [
    "",
  ],
  Array [
    "Name: test",
  ],
  Array [
    "Value: 12345",
  ],
  Array [
    "",
  ],
  Array [
    "[32mDONE![39m",
  ],
]
`;

exports[`Should deauth on 401 on creation 1`] = `
Array [
  Array [
    "[1mGenerating new deployment secret for:[22m",
    "http://localhost:8080",
  ],
  Array [
    "[31mError: authorization expired![39m",
    "Please, relogin and try again.",
  ],
]
`;

exports[`Should deauth on 401 on list 1`] = `
Array [
  Array [
    "[1mListing deployment secrets for:[22m",
    "http://localhost:8080",
  ],
  Array [
    "[31mError: authorization expired![39m",
    "Please, relogin and try again.",
  ],
]
`;

exports[`Should get secret value 1`] = `
Array [
  Array [
    "[1mGetting deployment secret for:[22m",
    "http://localhost:8080",
  ],
  Array [
    "[1mNew secret generated:[22m",
  ],
  Array [
    "",
  ],
  Array [
    "Name: undefined",
  ],
  Array [
    "Value: undefined",
  ],
  Array [
    "Date: 2/1/2018, 1:01:01 AM",
  ],
]
`;

exports[`Should list secrets 1`] = `
Array [
  Array [
    "[1mListing deployment secrets for:[22m",
    "http://localhost:8080",
  ],
  Array [
    "[1mGot saved secrets:[22m",
  ],
  Array [
    "",
  ],
  Array [
    "  > [32m@test[39m [90m[][39m",
  ],
]
`;

exports[`Should list zero secrets 1`] = `
Array [
  Array [
    "[1mListing deployment secrets for:[22m",
    "http://localhost:8080",
  ],
  Array [
    "[1mGot saved secrets:[22m",
  ],
  Array [
    "",
  ],
  Array [
    "  > No deployment secrets available!",
  ],
]
`;

exports[`Should remove secret 1`] = `
Array [
  Array [
    "[1mRemoving deployment secret for:[22m",
    "http://localhost:8080",
  ],
  Array [
    "[32mDeployment secret successfully removed![39m",
  ],
]
`;
